File: javabase/src/main/java/com/ellen/javabase/simpleapi/app/AppLifeListener.java
package com.ellen.javabase.simpleapi.app;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

public abstract class AppLifeListener implements Application.ActivityLifecycleCallbacks {

    private int lifeActivityCount = 0;
    private boolean isAppBack = false;
    //是否冷启动
    private boolean isColdLaunch = true;
    private List<Activity> activityList;

    @Override
    public void onActivityCreated(@NonNull Activity activity, @Nullable Bundle savedInstanceState) {
        if (lifeActivityCount == 0) {
            if(isColdLaunch) {
                onAppLaunch(true);
                isColdLaunch = false;
            }else {
                onAppLaunch(false);
            }
        }
        if (activityList == null) {
            activityList = new ArrayList<>();
        }
        activityList.add(activity);
    }

    @Override
    public void onActivityStarted(@NonNull Activity activity) {

    }

    @Override
    public void onActivityResumed(@NonNull Activity activity) {
        lifeActivityCount++;
        if (isAppBack) {
            //从后台切换至前台
            onSwitchFront();
            isAppBack = false;
        }
    }

    @Override
    public void onActivityPaused(@NonNull Activity activity) {

    }

    @Override
    public void onActivityStopped(@NonNull Activity activity) {
        lifeActivityCount--;
        if (lifeActivityCount == 0) {
            //切换至后台了
            isAppBack = true;
            onSwitchBack();
        }
    }

    @Override
    public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle outState) {

    }

    @Override
    public void onActivityDestroyed(@NonNull Activity activity) {
        activityList.remove(activity);
        if (activityList != null && activityList.size() == 0) {
            onCloseApp();

            //初始化
            lifeActivityCount = 0;
            isAppBack = false;
            activityList.clear();
        }
    }

    /**
     * 应用首次启动的时候调用
     */
    protected abstract void onAppLaunch(boolean isColdLaunch);

    /**
     * 切换至后台时调用
     */
    protected abstract void onSwitchBack();

    /**
     * 从后台切换至前台时调用
     */
    protected abstract void onSwitchFront();

    /**
     * 关闭应用时回调
     */
    protected abstract void onCloseApp();

    public void quitApp() {
        for (Activity activity : activityList) {
            activity.finish();
        }
    }
}

File: base/src/main/java/com/ellen/base/app/AppLifeListener.kt
package com.ellen.base.app

import android.app.Activity
import android.app.Application
import android.os.Bundle

abstract class AppLifeListener:Application.ActivityLifecycleCallbacks{

    private var lifeActivityCount = 0
    private var isAppBack = false
    //是否冷启动
    private var isColdLaunch = true
    //kotlin中ArrayList应该这么new
    private var activityList: MutableList<Activity> = ArrayList()

    override fun onActivityPaused(activity: Activity) {
    }

    override fun onActivityStarted(activity: Activity) {
    }

    override fun onActivityDestroyed(activity: Activity) {
        activityList.remove(activity)
        if(activityList.size == 0){
            onCloseApp()

            //进行初始化
            lifeActivityCount = 0
            isAppBack = false
            activityList = ArrayList()
        }
    }

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
    }

    override fun onActivityStopped(activity: Activity) {
        lifeActivityCount--
        if(lifeActivityCount == 0){
            onSwitchBack()
            isAppBack = true
        }
    }

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
        if(lifeActivityCount == 0){
            if(isColdLaunch) {
                //冷启动
                onAppLaunch(true)
                isColdLaunch = false
            }else{
                //热启动 & 温启动
                onAppLaunch(false)
            }
        }
        activityList.add(activity)
    }

    override fun onActivityResumed(activity: Activity) {
        lifeActivityCount++
        if(isAppBack){
            onSwitchFront()
            isAppBack = false
        }
    }

    /**
     * 只要应用启动就会调用
     * isColdLaunch:是否冷启动
     */
    protected abstract fun onAppLaunch(isColdLaunch: Boolean)

    /**
     * 切换至后台时调用
     */
    protected abstract fun onSwitchBack()

    /**
     * 从后台切换至前台时调用
     */
    protected abstract fun onSwitchFront()

    /**
     * 关闭应用时回调
     */
    protected abstract fun onCloseApp()
}